### PR TITLE
Replace hard-coded store name and total price fields in receipt entity.

### PR DESCRIPTION
**Pull request by basaam0**
- Send store name and total price from the upload form to the upload servlet
- Insert fields in Datastore entity
- Update upload servlet unit tests
- Add unit test for rounding total price

### CODE REVIEW

File: src/test/java/UploadReceiptServletTest.java [hunk 1]
```diff
@@ -205,6 +209,8 @@ public void doPostUploadsReceiptToDatastoreDevServer() throws IOException {
     when(blobInfoFactory.loadBlobInfo(BLOB_KEY)).thenReturn(blobInfo);
 
     when(request.getParameter("label")).thenReturn(LABEL);
```
Comment (emilyvera) [inline src/test/java/UploadReceiptServletTest.java:211]: I mentioned this in another PR but since a lot of this setup code for each test is repeated, maybe it can be put into a helper function

Comment (basaam0) [reply src/test/java/UploadReceiptServletTest.java]: I've added some helper functions here and in PR #27.

File: src/test/java/UploadReceiptServletTest.java [hunk 2]
```diff
@@ -93,6 +93,8 @@
   private static final String LABEL = "Label";
   private static final Text RAW_TEXT = new Text("raw text");
   private static final double PRICE = 5.89;
+  private static final double PRICE_THREE_DECIMAL_PLACES = 17.236;
+  private static final double PRICE_THREE_DECIMAL_PLACES_ROUNDED = 17.24;
```
Comment (bryyeh) [inline src/test/java/UploadReceiptServletTest.java:97]: If the variable isn't going to be used by more than one test, I prefer to place it inside the test method. This way the list of class variables doesn't become long and unwieldy.

File: src/test/java/UploadReceiptServletTest.java [hunk 3]
```diff
@@ -329,4 +319,60 @@ public void doPostThrowsIfReceiptAnalysisFails() throws IOException {
...
+    Assert.assertEquals(receipt.getProperty("rawText"), RAW_TEXT);
+    Assert.assertEquals(receipt.getProperty("blobKey"), BLOB_KEY);
+    Assert.assertEquals(receipt.getProperty("timestamp"), timestamp);
+    Assert.assertEquals(receipt.getProperty("label"), LABEL);
+  }
+
+  /**
+   * Stubs the request with the given label, store, and price parameters.
+   */
+  private void stubRequestBody(
```
Comment (bryyeh) [inline src/test/java/UploadReceiptServletTest.java:361]: I like the helper methods :)

**Comment by basaam0** [comment]
> Is there a way to give the user feedback that the total price should only be 2 decimal places in the input box? This way the price isn't unexpectedly rounded for them.

The input box already restricts the user to 2 decimal places. The rounding is meant to be a safeguard on the back end and would not have any effect if the user submits from our front end.

File: src/main/java/servlets/UploadReceiptServlet.java [hunk 4]
```diff
@@ -216,6 +220,13 @@ private static boolean isValidFilename(String filename) {
     return validFilename.matcher(filename).matches();
   }
 
+  /**
+   * Converts a price string into a double rounded to 2 decimal places.
+   */
+  private static double roundPrice(String price) {
+    return Math.round(Double.parseDouble(price) * 100.0) / 100.0;
```
Comment (bryyeh) [inline src/main/java/servlets/UploadReceiptServlet.java:227]: Do we also need a check to make sure the price is not negative?

File: src/test/java/UploadReceiptServletTest.java [hunk 5]
```diff
@@ -394,6 +381,42 @@ public void doPostThrowsIfReceiptAnalysisFails() throws IOException {
     verify(blobstoreService).delete(BLOB_KEY);
   }
 
+  @Test
+  public void doPostRoundPrice() throws IOException {
+    helper.setEnvIsLoggedIn(true);
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
```
Comment (acheng1) [inline src/test/java/UploadReceiptServletTest.java:390]: Why is this needed for this test but not for doPostUploadsReceiptToDatastoreLiveServer?

File: src/test/java/UploadReceiptServletTest.java [hunk 6]
```diff
@@ -394,6 +381,42 @@ public void doPostThrowsIfReceiptAnalysisFails() throws IOException {
...
+    PreparedQuery results = datastore.prepare(query);
+    Entity receipt = results.asSingleEntity();
+
+    Assert.assertEquals(receipt.getProperty("imageUrl"), IMAGE_URL);
+    Assert.assertEquals(receipt.getProperty("price"), roundedPrice);
+    Assert.assertEquals(receipt.getProperty("store"), STORE);
+    Assert.assertEquals(receipt.getProperty("rawText"), RAW_TEXT);
+    Assert.assertEquals(receipt.getProperty("blobKey"), BLOB_KEY);
+    Assert.assertEquals(receipt.getProperty("timestamp"), PAST_TIMESTAMP);
+    Assert.assertEquals(receipt.getProperty("label"), LABEL);
```
Comment (acheng1) [inline src/test/java/UploadReceiptServletTest.java:417]: I would probably only assert on the price since the others are redundant with the assertions in doPostUploadsReceiptToDatastoreLiveServer. Generally we don't want to assert more than is required to verify a behavior.
